src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoNullable.cs

```diff
@@ -32,54 +32,56 @@ internal sealed class JsonPropertyInfoNullable<TClass, TProperty>
         public override void Read(JsonTokenType tokenType, JsonSerializerOptions options, ref ReadStack state, ref Utf8JsonReader reader)
         {
             Debug.Assert(ElementClassInfo == null);
+            Debug.Assert(ShouldDeserialize);
 
-            if (ShouldDeserialize)
+            if (ValueConverter != null)
```

Comment:
Can we invert this?

Something like:
```C#
public override void Read(JsonTokenType tokenType, JsonSerializerOptions options, ref ReadStack state, ref Utf8JsonReader reader)
{
    Debug.Assert(ElementClassInfo == null);
    Debug.Assert(ShouldDeserialize);

    if (ValueConverter == null || !(ValueConverter.TryRead(s_underlyingType, ref reader, out TProperty value)))
    {
        ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(RuntimePropertyType, reader, state.PropertyPath);
    }

    if (state.Current.ReturnValue == null)
    {
        state.Current.ReturnValue = value;
    }
    else
    {
        Set((TClass)state.Current.ReturnValue, value);
    }
}
```

Reply:
> The code above actually doesn't compile

Sure. Having the local be defined outside is fine, to make it compile. But it does get rid of the super deep nesting (with return in the middle).

Reply:
I can remove one level easily by combining the two `ifs`.

However another reason it doesn't compile is that the compiler will think `value` is unassigned.